include/seqan3/alphabet/concept.hpp

```diff
@@ -16,22 +16,936 @@
+{
+
+//!\brief Call `assign_rank_to()` on the argument, but only consider ADL.
+//!\ingroup alphabet
+template <typename alph_t>
+//!\cond
+    requires requires { assign_rank_to(std::declval<alphabet_rank_t<alph_t>>(), std::declval<alph_t &>()); }
+//!\endcond
+constexpr decltype(auto) assign_rank_to_(alphabet_rank_t<alph_t> const r, alph_t & a)
+    noexcept(noexcept(assign_rank_to(std::declval<alphabet_rank_t<alph_t>>(), std::declval<alph_t &>())))
+{
+    return assign_rank_to(r, a);
+}
+
+} // namespace seqan3::detail::adl::only
+
+namespace seqan3::detail
+{
+
+// Auxiliary concepts
+//!\cond
+template <typename alph_t>
+SEQAN3_CONCEPT has_assign_rank_to_adl_overload = requires
+{
+    { adl::only::assign_rank_to_(std::declval<alphabet_rank_t<alph_t>>(), std::declval<alph_t &>()) };
+};
+
+template <typename alph_t>
+SEQAN3_CONCEPT has_assign_rank_to_adaptation_overload = requires
+{
+    { ::seqan3::adaptation::assign_rank_to(std::declval<alphabet_rank_t<alph_t>>(), std::declval<alph_t &>()) };
+};
+
+template <typename alph_t>
+SEQAN3_CONCEPT has_assign_rank_member = requires
+{
+    { std::declval<alph_t &>().assign_rank(std::declval<alphabet_rank_t<alph_t>>()) };
+};
+//!\endcond
+
+//!\brief Functor definition for seqan3::assign_rank_to.
+//!\ingroup alphabet
+struct assign_rank_to_fn
+{
+    //!\brief Operator definition for lvalues.
+    template <typename alph_t>
+    //!\cond
+        requires has_assign_rank_to_adl_overload<alph_t>        ||
+                 has_assign_rank_to_adaptation_overload<alph_t> ||
+                 has_assign_rank_member<alph_t>
+    //!\endcond
+    constexpr decltype(auto) operator()(alphabet_rank_t<alph_t> const r, alph_t & a) const noexcept
+    {
+        if constexpr (has_assign_rank_to_adl_overload<alph_t>)
+        {
+            static_assert(noexcept(adl::only::assign_rank_to_(r, a)),
+                          "Only overloads that are marked noexcept are picked up by seqan3::assign_rank_to().");
+            return adl::only::assign_rank_to_(r, a);
+        }
+        else if constexpr (has_assign_rank_to_adaptation_overload<alph_t>)
+        {
+            static_assert(noexcept(seqan3::adaptation::assign_rank_to(r, a)),
+                          "Only overloads that are marked noexcept are picked up by seqan3::assign_rank_to().");
+            return seqan3::adaptation::assign_rank_to(r, a);
+        }
+        else
+        {
+            static_assert(noexcept(a.assign_rank(r)),
+                          "Only members that are marked noexcept are picked up by seqan3::assign_rank_to().");
+            return a.assign_rank(r);
+        }
+    }
+
+    //!\brief Operator definition for rvalues.
+    template <typename alph_t>
+    //!\cond
+        requires (has_assign_rank_to_adl_overload<alph_t>        ||
+                  has_assign_rank_to_adaptation_overload<alph_t> ||
+                  has_assign_rank_member<alph_t>) &&
+                  (!std::is_lvalue_reference_v<alph_t>)
```

Comment:
-> `std::is_rvalue_reference<alph_t>` then you don't the negation nor the paranthesis?

Reply:
That won't work. If you pass a temporary of `T`, the type of `alph_t` will be `T`, not `T &&`. Only lvalue references are "part of the type" because they need to be for reference collapsing.